api/extension-guides/language-model-tutorial.md

```diff
@@ -9,13 +9,324 @@ MetaDescription: Tutorial that walks you through creating a VS Code extension th
+ while (currentLine < endLine) {
+  code += `${currentLine + 1}: ${textEditor.document.lineAt(currentLine).text} \n`;
+  currentLine++;
+ }
+ return code;
+}
+```
+
+This code uses the `visibleRanges` property of the TextEditor to get the lines that are currently visible in the editor. It then loops through these lines and adds each one to a string, along with the line number. Finally, it returns the string that contains the code with line numbers.
+
+Now we can call this method from the `code-tutor.annotate` command. Modify the implementation of the command so that it looks like this:
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getCodeWithLineNumbers(textEditor);
+
+});
+```
+
+The next step is to call the GitHub Copilot language model. To do this, we first need to specify which chat model we want to use.
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getCodeWithLineNumbers(textEditor);
+
+  // select the 4o chat model
+  let [model] = await vscode.lm.selectChatModels({
+    vendor: 'copilot',
+    family: 'gpt-4o',
+  });
+});
+```
+
+Now we need to send the code to the language model along with a prompt that tells the model to generate annotations for the code. Add the following code to the top of the file directly under the imports.
+
+```ts
+const ANNOTATION_PROMPT = `You are a code tutor who helps students learn how to write better code. Your job is to evaluate a block of code that the user gives you. The user is writing You will then annotate any lines that could be improved with a brief suggestion and the reason why you are making that suggestion. Only make suggestions when you feel the severity is enough that it will impact the readibility and maintainability of the code. Be friendly with your suggestions and remember that these are students so they need gentle guidance. Format each suggestion as a single JSON object. It is not necessary to wrap your response in triple backticks. Here is an example of what your response should look like:
+
+{ "line": 1, "suggestion": "I think you should use a for loop instead of a while loop. A for loop is more concise and easier to read." }{ "line": 12, "suggestion": "I think you should use a for loop instead of a while loop. A for loop is more concise and easier to read." }
+`;
+```
+
+This is a special prompt that instructs the language model on how to generate annotations. It also includes examples for how the model should format its response. This is what allows to to know what format the response will be in, so that we can parse it and display it as annotations. Now we can pass the prompt and the code with line numbers to the model.
```

Comment:
```suggestion
This is a special prompt that instructs the language model on how to generate annotations. It also includes examples for how the model should format its response. This is what allows it to know what format the response will be in, so that we can parse it and display it as annotations. Now we can pass the prompt and the code with line numbers to the model.
```